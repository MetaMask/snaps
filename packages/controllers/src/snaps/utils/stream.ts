import { Readable, Writable } from 'stream';
import { ReadableWebToNodeStream } from 'readable-web-to-node-stream';
import { extract as tarExtract } from 'tar-stream';
import concat from 'concat-stream';
import { isObject } from '@metamask/utils';
import {
  SnapManifest,
  NpmSnapFileNames,
  UnvalidatedSnapFiles,
} from '@metamask/snap-utils';

// The paths of files within npm tarballs appear to always be prefixed with
// "package/".
const NPM_TARBALL_PATH_PREFIX = /^package\//u;

/**
 * Strips the leading `./` from a string, or does nothing if no string is
 * provided.
 *
 * @param pathString - The path string to normalize.
 * @returns The specified path without a `./` prefix, or `undefined` if no
 * string was provided.
 */
export function stripDotSlash(pathString?: string): string | undefined {
  return pathString?.replace(/^\.\//u, '');
}

/**
 * Converts a {@link ReadableStream} to a Node.js {@link Readable}
 * stream. Returns the stream directly if it is already a Node.js stream.
 * We can't use the native Web {@link ReadableStream} directly because the
 * other stream libraries we use expect Node.js streams.
 *
 * @param stream - The stream to convert.
 * @returns The given stream as a Node.js Readable stream.
 */
export function getNodeStream(stream: ReadableStream): Readable {
  if (typeof stream.getReader !== 'function') {
    return stream as unknown as Readable;
  }

  return new ReadableWebToNodeStream(stream);
}

/**
 * Creates a `tar-stream` that will get the necessary files from an npm Snap
 * package tarball (`.tgz` file).
 *
 * @param snapFiles - An object to write target file contents to.
 * @returns The {@link Writable} tarball extraction stream.
 */
export function createTarballExtractionStream(
  snapFiles: UnvalidatedSnapFiles,
): Writable {
  // `tar-stream` is pretty old-school, so we create it first and then
  // instrument it by adding event listeners.
  const extractStream = tarExtract();

  // `tar-stream` reads every file in the tarball serially. We already know
  // where to look for package.json and the Snap manifest, but we don't know
  // where the source code is. Therefore, we cache the contents of each .js
  // file in the tarball and pick out the correct one when the stream has ended.
  const jsFileCache = new Map<string, Buffer>();

  // "entry" is fired for every discreet entity in the tarball. This includes
  // files and folders.
  extractStream.on('entry', (header, entryStream, next) => {
    const { name: headerName, type: headerType } = header;
    if (headerType === 'file') {
      // The name is a path if the header type is "file".
      const filePath = headerName.replace(NPM_TARBALL_PATH_PREFIX, '');

      // Note the use of `concat-stream` since the data for each file may be
      // chunked.
      if (filePath === NpmSnapFileNames.PackageJson) {
        return entryStream.pipe(
          concat((data) => {
            try {
              snapFiles.packageJson = JSON.parse(data.toString());
            } catch (_error) {
              return extractStream.destroy(
                new Error(`Failed to parse "${NpmSnapFileNames.PackageJson}".`),
              );
            }

            return next();
          }),
        );
      } else if (filePath === NpmSnapFileNames.Manifest) {
        return entryStream.pipe(
          concat((data) => {
            try {
              snapFiles.manifest = JSON.parse(data.toString());
            } catch (_error) {
              return extractStream.destroy(
                new Error(`Failed to parse "${NpmSnapFileNames.Manifest}".`),
              );
            }

            return next();
          }),
        );
      } else if (/\w+\.(?:js|svg)$/u.test(filePath)) {
        return entryStream.pipe(
          concat((data) => {
            jsFileCache.set(filePath, data);
            return next();
          }),
        );
      }
    }

    // If we get here, the entry is not a file, and we want to ignore. The entry
    // stream must be drained, or the extractStream will stop reading. This is
    // effectively a no-op for the current entry.
    entryStream.on('end', () => next());
    return entryStream.resume();
  });

  // Once we've read the entire tarball, attempt to grab the bundle file
  // contents from the .js file cache.
  extractStream.once('finish', () => {
    if (isObject(snapFiles.manifest)) {
      /* istanbul ignore next: optional chaining */
      const { filePath: _bundlePath, iconPath: _iconPath } =
        (snapFiles.manifest as unknown as Partial<SnapManifest>).source
          ?.location?.npm ?? {};

      const bundlePath = stripDotSlash(_bundlePath);
      const iconPath = stripDotSlash(_iconPath);

      if (bundlePath) {
        snapFiles.sourceCode = jsFileCache.get(bundlePath)?.toString('utf8');
      }

      if (typeof iconPath === 'string' && iconPath.endsWith('.svg')) {
        snapFiles.svgIcon = jsFileCache.get(iconPath)?.toString('utf8');
      }
    }
    jsFileCache.clear();
  });

  return extractStream;
}
