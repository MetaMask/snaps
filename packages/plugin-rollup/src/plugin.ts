import { postProcessBundle, PostProcessOptions } from '@metamask/snap-utils';
import { Plugin } from 'rollup';

export type Options = PostProcessOptions;

/**
 * Creates a Snaps Rollup plugin instance.
 *
 * @param options - The plugin options.
 * @param options.stripComments - Whether to strip comments. Defaults to `true`.
 * @param options.transformHtmlComments - Whether to transform HTML comments.
 * Defaults to `true`.
 * @returns The Rollup plugin object.
 */
export default function snaps(options: Partial<Options> = {}): Plugin {
  return {
    name: '@metamask/rollup-plugin-snaps',
    renderChunk(code) {
      return postProcessBundle(code, options);
    },
  };
}
