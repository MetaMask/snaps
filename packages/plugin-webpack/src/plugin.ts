import { postProcessBundle, PostProcessOptions } from '@metamask/snap-utils';
import { Compiler } from 'webpack';
import { RawSource } from 'webpack-sources';

const PLUGIN_NAME = 'SnapsWebpackPlugin';

export type Options = PostProcessOptions;

export default class SnapsWebpackPlugin {
  public readonly options: Partial<Options>;

  /**
   * Construct an instance of the plugin.
   *
   * @param options - The post-process options.
   * @param options.stripComments - Whether to strip comments. Defaults to `true`.
   * @param options.transformHtmlComments - Whether to transform HTML comments.
   * Defaults to `true`.
   */
  constructor(options: Partial<Options> = {}) {
    this.options = options;
  }

  /**
   * Apply the plugin to the Webpack compiler. Hooks into the `processAssets`
   * stage to process the bundle.
   *
   * @param compiler - The Webpack compiler.
   */
  apply(compiler: Compiler) {
    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
      compilation.hooks.processAssets.tap(PLUGIN_NAME, (assets) => {
        Object.keys(assets).forEach((assetName) => {
          const asset = assets[assetName];
          const processed = postProcessBundle(
            asset.source() as string,
            this.options,
          );

          if (processed) {
            // For some reason the type of `RawSource` is not compatible with Webpack's own
            // `Source`, but works fine when casting it to `any`.
            compilation.updateAsset(assetName, new RawSource(processed) as any);
          }
        });
      });
    });
  }
}
