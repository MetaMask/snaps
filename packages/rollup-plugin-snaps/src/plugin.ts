import { promises as fs } from 'fs';
import pathUtils from 'path';
import {
  checkManifest,
  evalBundle,
  postProcessBundle,
  PostProcessOptions,
} from '@metamask/snap-utils';
import { Plugin, SourceMapInput } from 'rollup';

type PluginOptions = {
  eval?: boolean;
  manifestPath?: string;
  writeManifest?: boolean;
};

export type Options = PluginOptions &
  Omit<PostProcessOptions, 'sourceMap' | 'inputSourceMap'>;

/**
 * Creates a Snaps Rollup plugin instance.
 *
 * @param options - The plugin options.
 * @param options.stripComments - Whether to strip comments. Defaults to `true`.
 * @param options.eval - Whether to evaluate the bundle to test SES
 * compatibility. Defaults to `true`.
 * @param options.manifestPath - The path to the manifest file. If provided,
 * the manifest will be validated. Defaults to
 * `process.cwd() + '/snap.manifest.json'`.
 * @param options.writeManifest - Whether to fix the manifest. Defaults to
 * `true`.
 * @returns The Rollup plugin object.
 */
export default function snaps(options?: Partial<Options>): Plugin {
  const defaultOptions = {
    eval: true,
    manifestPath: pathUtils.join(process.cwd(), 'snap.manifest.json'),
    writeManifest: true,
    ...options,
  };

  return {
    name: '@metamask/rollup-plugin-snaps',

    renderChunk(code: string): { code: string; map?: SourceMapInput } | null {
      // Rollup internally merges the new source map with the old one, so there
      // is no need to pass the current source map to the function.
      const result = postProcessBundle(code, {
        ...defaultOptions,
        sourceMap: true,
      });

      return { code: result.code, map: result.sourceMap };
    },

    async writeBundle(output): Promise<void> {
      if (!output.file) {
        this.warn('No output file specified, skipping bundle validation.');
        return;
      }

      if (defaultOptions.eval) {
        await evalBundle(output.file).catch((error) => {
          this.error(error);
        });
      }

      if (defaultOptions.manifestPath) {
        const { errors, warnings } = await checkManifest(
          pathUtils.dirname(defaultOptions.manifestPath),
          defaultOptions.writeManifest,
          await fs.readFile(output.file, 'utf8'),
        );

        if (!defaultOptions.writeManifest && errors.length > 0) {
          this.error(
            `Manifest Error: The manifest is invalid.\n${errors.join('\n')}`,
          );
        }

        if (warnings.length > 0) {
          this.warn(
            `Manifest Warning: Validation of snap.manifest.json completed with warnings.\n${warnings.join(
              '\n',
            )}`,
          );
        }
      }
    },
  };
}
