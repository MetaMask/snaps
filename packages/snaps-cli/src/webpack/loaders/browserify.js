// This file is used for end-to-end tests only.

module.exports = require('./browserify.ts');
