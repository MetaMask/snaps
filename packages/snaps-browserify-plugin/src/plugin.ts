import type { PostProcessOptions } from '@metamask/snaps-utils';
import {
  checkManifest,
  evalBundle,
  logWarning,
  postProcessBundle,
  useTemporaryFile,
} from '@metamask/snaps-utils';
import type { BrowserifyObject } from 'browserify';
import { fromSource } from 'convert-source-map';
import pathUtils from 'path';
import type { TransformCallback } from 'stream';
import { Transform } from 'stream';

type PluginOptions = {
  eval?: boolean;
  manifestPath?: string;
  writeManifest?: boolean;
};

export type Options = PluginOptions &
  Omit<PostProcessOptions, 'sourceMap' | 'inputSourceMap'>;

/**
 * Run eval on the processed bundle and fix the manifest, if configured.
 *
 * @param options - The plugin options.
 * @param code - The code to eval, if the eval option is enabled.
 */
async function postBundle(options: Partial<Options>, code: string) {
  if (options.eval) {
    await useTemporaryFile('snaps-bundle.js', code, (path) => evalBundle(path));
  }

  if (options.manifestPath) {
    const { errors, warnings } = await checkManifest(
      pathUtils.dirname(options.manifestPath),
      options.writeManifest,
      code,
    );

    if (!options.writeManifest && errors.length > 0) {
      throw new Error(
        `Manifest Error: The manifest is invalid.\n${errors.join('\n')}`,
      );
    }

    if (warnings.length > 0) {
      logWarning(
        'Manifest Warning: Validation of snap.manifest.json completed with warnings.',
      );

      warnings.forEach((warning) => logWarning(`Manifest Warning: ${warning}`));
    }
  }
}

/**
 * A transform stream which can be used in the Browserify pipeline. It accepts a
 * string input, which is post-processed and pushed to the output stream.
 */
export class SnapsBrowserifyTransform extends Transform {
  readonly #data: Buffer[] = [];

  readonly #options: Partial<Options>;

  /**
   * Construct an instance of the transform stream.
   *
   * @param options - The post-processing options.
   * @param options.stripComments - Whether to strip comments. Defaults to `true`.
   * @param options.transformHtmlComments - Whether to transform HTML comments.
   * Defaults to `true`.
   */
  constructor(options: Partial<Options> = {}) {
    super();
    this.#options = { ...options };
  }

  /**
   * Takes a chunk of data and pushes it into an internal array, for later
   * processing.
   *
   * @param chunk - The chunk of data to transform.
   * @param _encoding - The encoding of the chunk.
   * @param callback - The callback to call when the chunk is processed.
   */
  _transform(
    chunk: Buffer,
    _encoding: BufferEncoding,
    callback: TransformCallback,
  ) {
    // Collects all the chunks into an array.
    this.#data.push(chunk);
    callback();
  }

  /**
   * Takes the internal array of chunks and processes them. The processed code
   * is pushed to the output stream.
   *
   * @param callback - The callback to call when the stream is finished.
   */
  _flush(callback: TransformCallback) {
    // Merges all the chunks into a single string and processes it.
    const code = Buffer.concat(this.#data).toString('utf-8');

    // Browserify uses inline source maps, so we attempt to read it here, and
    // convert it to an object.
    const inputSourceMap = fromSource(code)?.toObject() ?? undefined;

    const result = postProcessBundle(code, {
      ...this.#options,
      sourceMap: Boolean(inputSourceMap) && 'inline',
      inputSourceMap,
    });

    if (result.warnings.length > 0) {
      logWarning(
        `Bundle Warning: Processing of the Snap bundle completed with warnings.\n${result.warnings.join(
          '\n',
        )}`,
      );
    }

    postBundle(this.#options, result.code)
      .catch((error) => {
        callback(error);
      })
      .finally(() => {
        this.push(result.code);
        callback();
      });
  }
}

/**
 * The Browserify plugin function. Can be passed to the Browserify `plugin`
 * function, or used by simply passing the package name to `plugin`.
 *
 * @param browserifyInstance - The Browserify instance.
 * @param options - The plugin options.
 * @param options.stripComments - Whether to strip comments. Defaults to `true`.
 * @param options.eval - Whether to evaluate the bundle to test SES
 * compatibility. Defaults to `true`.
 * @param options.manifestPath - The path to the manifest file. If provided,
 * the manifest will be validated. Defaults to
 * `process.cwd() + '/snap.manifest.json'`.
 * @param options.writeManifest - Whether to fix the manifest.
 * Defaults to `true`.
 */
export default function plugin(
  browserifyInstance: BrowserifyObject,
  options?: Partial<Options>,
): void {
  const defaultOptions = {
    eval: true,
    manifestPath: pathUtils.join(process.cwd(), 'snap.manifest.json'),
    writeManifest: true,
    ...options,
  };

  // Pushes the transform stream at the end of Browserify's pipeline. This
  // ensures that the transform is run on the entire bundle.
  browserifyInstance.pipeline.push(
    new SnapsBrowserifyTransform(defaultOptions),
  );
}
