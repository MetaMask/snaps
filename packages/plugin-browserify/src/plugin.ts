import { Transform, TransformCallback } from 'stream';
import { promises as fs } from 'fs';
import os from 'os';
import pathUtils from 'path';
import { BrowserifyObject } from 'browserify';
import {
  checkManifest,
  evalBundle,
  postProcessBundle,
  PostProcessOptions,
} from '@metamask/snap-utils';
import { fromSource } from 'convert-source-map';

const TEMP_BUNDLE_PATH = pathUtils.join(os.tmpdir(), 'snaps-bundle.js');

type PluginOptions = {
  eval?: boolean;
  manifestPath?: string;
  writeManifest?: boolean;
};

export type Options = PluginOptions &
  Omit<PostProcessOptions, 'sourceMap' | 'inputSourceMap'>;

/**
 * Run eval on the processed bundle and fix the manifest, if configured.
 *
 * @param options - The plugin options.
 * @param code - The code to eval, if the eval option is enabled.
 */
async function postBundle(options: Partial<Options>, code: string) {
  if (options.eval) {
    await fs.mkdir(pathUtils.dirname(TEMP_BUNDLE_PATH), { recursive: true });
    await fs.writeFile(TEMP_BUNDLE_PATH, code);

    await evalBundle(TEMP_BUNDLE_PATH)
      .catch((error) => {
        throw new Error(`Snap evaluation error: ${error.toString()}`);
      })
      .finally(() => fs.unlink(TEMP_BUNDLE_PATH));
  }

  if (options.manifestPath) {
    const { errors, warnings } = await checkManifest(
      pathUtils.dirname(options.manifestPath),
      options.writeManifest,
      code,
    );

    if (!options.writeManifest && errors.length > 0) {
      throw new Error(
        `Manifest Error: The manifest is invalid.\n${errors.join('\n')}`,
      );
    }

    if (warnings.length > 0) {
      console.log(
        'Manifest Warning: Validation of snap.manifest.json completed with warnings.',
      );

      warnings.forEach((warning) =>
        console.log(`Manifest Warning: ${warning}`),
      );
    }
  }
}

/**
 * A transform stream which can be used in the Browserify pipeline. It accepts a
 * string input, which is post-processed and pushed to the output stream.
 */
export class SnapsBrowserifyTransform extends Transform {
  readonly #data: Buffer[] = [];

  readonly #options: Partial<Options>;

  /**
   * Construct an instance of the transform stream.
   *
   * @param options - The post-processing options.
   * @param options.stripComments - Whether to strip comments. Defaults to `true`.
   * @param options.transformHtmlComments - Whether to transform HTML comments.
   * Defaults to `true`.
   */
  constructor(options: Partial<Options> = {}) {
    super();
    this.#options = { ...options };
  }

  /**
   * Takes a chunk of data and pushes it into an internal array, for later
   * processing.
   *
   * @param chunk - The chunk of data to transform.
   * @param _encoding - The encoding of the chunk.
   * @param callback - The callback to call when the chunk is processed.
   */
  _transform(
    chunk: Buffer,
    _encoding: BufferEncoding,
    callback: TransformCallback,
  ) {
    // Collects all the chunks into an array.
    this.#data.push(chunk);
    callback();
  }

  /**
   * Takes the internal array of chunks and processes them. The processed code
   * is pushed to the output stream.
   *
   * @param callback - The callback to call when the stream is finished.
   */
  _flush(callback: TransformCallback) {
    // Merges all the chunks into a single string and processes it.
    const code = Buffer.concat(this.#data).toString('utf-8');

    // Browserify uses inline source maps, so we attempt to read it here, and
    // convert it to an object.
    const inputSourceMap = fromSource(code)?.toObject() ?? undefined;

    const result = postProcessBundle(code, {
      ...this.#options,
      sourceMap: Boolean(inputSourceMap) && 'inline',
      inputSourceMap,
    });

    postBundle(this.#options, result.code)
      .catch((error) => {
        callback(error);
      })
      .finally(() => {
        this.push(result.code);
        callback();
      });
  }
}

/**
 * The Browserify plugin function. Can be passed to the Browserify `plugin`
 * function, or used by simply passing the package name to `plugin`.
 *
 * @param browserifyInstance - The Browserify instance.
 * @param options - The plugin options.
 * @param options.stripComments - Whether to strip comments. Defaults to `true`.
 * @param options.eval - Whether to evaluate the bundle to test SES
 * compatibility. Defaults to `true`.
 * @param options.manifestPath - The path to the manifest file. If provided,
 * the manifest will be validated. Defaults to
 * `process.cwd() + '/snap.manifest.json'`.
 * @param options.writeManifest - Whether to fix the manifest.
 * Defaults to `true`.
 */
export default function plugin(
  browserifyInstance: BrowserifyObject,
  options?: Partial<Options>,
): void {
  const defaultOptions = {
    eval: true,
    manifestPath: pathUtils.join(process.cwd(), 'snap.manifest.json'),
    writeManifest: true,
    ...options,
  };

  // Pushes the transform stream at the end of Browserify's pipeline. This
  // ensures that the transform is run on the entire bundle.
  browserifyInstance.pipeline.push(
    new SnapsBrowserifyTransform(defaultOptions),
  );
}
