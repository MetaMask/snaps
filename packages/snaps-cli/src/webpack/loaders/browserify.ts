import browserify from 'browserify';
import { Readable } from 'stream';
import type { LoaderDefinitionFunction } from 'webpack';

import { TranspilationModes } from '../../builders';
import type { LegacyOptions } from '../../config';
import { processDependencies } from '../../utils';
import { getBrowserslistTargets } from '../utils';

/**
 * A Browserify loader for Webpack. This exists for backwards compatibility with
 * the legacy configuration format, in order to support the `bundlerCustomizer`
 * function.
 *
 * When this loader is used, the input file will be processed by Browserify, and
 * written to disk by Webpack. Most processing will be handled by Browserify, so
 * there are no benefits like tree-shaking.
 *
 * @param content - The input file contents as a string.
 * @param sourceMap - The source map of the input file.
 */
const loader: LoaderDefinitionFunction<LegacyOptions> = async function (
  content,
  sourceMap,
) {
  const config = this.getOptions();

  const { transpilationMode } = config;

  const bundler = browserify({
    extensions: ['.js', '.ts'],
    debug: Boolean(sourceMap),
    standalone: '<snap>',
  });

  if (transpilationMode !== TranspilationModes.None) {
    const babelifyOptions = processDependencies(config);

    // We need to statically import all Browserify transforms, and all Babel
    // presets and plugins, and calling `require` is the sanest way to do that.
    /* eslint-disable @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires, n/global-require */
    bundler.transform(require('babelify'), {
      global: transpilationMode === TranspilationModes.LocalAndDeps,
      extensions: ['.js', '.ts'],
      presets: [
        require('@babel/preset-typescript'),
        [
          require('@babel/preset-env'),
          {
            targets: {
              browsers: await getBrowserslistTargets(),
            },
          },
        ],
      ],
      plugins: [
        require('@babel/plugin-transform-runtime'),
        require('@babel/plugin-proposal-class-properties'),
        require('@babel/plugin-proposal-private-methods'),
        require('@babel/plugin-proposal-class-static-block'),
        require('@babel/plugin-proposal-private-property-in-object'),
      ],
      ...(babelifyOptions as any),
    });
    /* eslint-enable @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires, n/global-require */
  }

  config.bundlerCustomizer?.(bundler);

  // Browserify doesn't accept a string as an entry point, so we need to convert
  // it to a stream.
  const stream = new Readable();
  stream.push(content);
  stream.push(null);

  bundler.add(stream, {
    file: this.resourcePath,
  });

  return new Promise<Buffer>((resolve, reject) => {
    bundler.bundle((bundleError, buffer: Buffer) => {
      if (bundleError) {
        reject(bundleError);
        return;
      }

      // Browserify inlines the source map, so we just pass the output buffer back
      // to Webpack.
      resolve(buffer);
    });
  });
};

export default loader;
