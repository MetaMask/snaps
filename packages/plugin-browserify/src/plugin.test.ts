import { Readable } from 'stream';
import browserify from 'browserify';
import concat from 'concat-stream';
import plugin, { SnapsBrowserifyTransform } from './plugin';

/**
 * Takes a string value and pushes it to a readable stream. Used for testing
 * purposes.
 *
 * @param value - The value to push to a readable stream.
 * @returns The readable stream containing the value.
 */
const toStream = (value: string) => {
  const readable = new Readable();
  readable.push(value);
  readable.push(null);

  return readable;
};

describe('SnapsBrowserifyTransform', () => {
  it('processes the data', async () => {
    const transform = new SnapsBrowserifyTransform({});
    const stream = toStream(' foo bar ');

    const result = await new Promise((resolve) => {
      const concatStream = concat((value) => resolve(value.toString('utf-8')));

      stream.pipe(transform).pipe(concatStream);
    });

    expect(result).toBe('foo bar');
  });

  it('works without options', async () => {
    const transform = new SnapsBrowserifyTransform();
    const stream = toStream(' foo bar ');

    const result = await new Promise((resolve) => {
      const concatStream = concat((value) => resolve(value.toString('utf-8')));

      stream.pipe(transform).pipe(concatStream);
    });

    expect(result).toBe('foo bar');
  });
});

describe('plugin', () => {
  it('processes files using Browserify', async () => {
    const value = toStream(`const foo = 'bar';`);

    const result = await new Promise((resolve, reject) => {
      const bundler = browserify();

      bundler.plugin(plugin);
      bundler.add(value);

      bundler.bundle((error, src) => {
        if (error) {
          return reject(error);
        }

        return resolve(src.toString('utf-8'));
      });
    });

    expect(result).toContain(`const foo = 'bar';`);
  });

  it('applies a transform', async () => {
    const value = toStream(`
      // foo bar
      /* baz qux */
      const foo = 'bar';
    `);

    const result = await new Promise((resolve, reject) => {
      const bundler = browserify();

      bundler.add(value);
      bundler.plugin(plugin);

      bundler.bundle((error, src) => {
        if (error) {
          return reject(error);
        }

        return resolve(src.toString('utf-8'));
      });
    });

    expect(result).not.toContain(`// foo bar`);
    expect(result).not.toContain(`/* baz qux */`);
  });

  it('forwards the options', async () => {
    const value = toStream(`
      // foo bar
      /* baz qux */
      const foo = 'bar';
    `);

    const result = await new Promise((resolve, reject) => {
      const bundler = browserify();

      bundler.plugin(plugin, { stripComments: false });
      bundler.add(value);

      bundler.bundle((error, src) => {
        if (error) {
          return reject(error);
        }

        return resolve(src.toString('utf-8'));
      });
    });

    expect(result).toContain(`// foo bar`);
    expect(result).toContain(`/* baz qux */`);
  });
});
