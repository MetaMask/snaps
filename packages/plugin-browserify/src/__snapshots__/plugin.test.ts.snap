// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin applies a transform 1`] = `
"(function () {
  function r(e, n, t) {
    function o(i, f) {
      if (!n[i]) {
        if (!e[i]) {
          var c = "function" == typeof require && require;
          if (!f && c) return c(i, !0);
          if (u) return u(i, !0);
          var a = new Error("Cannot find module '" + i + "'");
          throw a.code = "MODULE_NOT_FOUND", a;
        }

        var p = n[i] = {
          exports: {}
        };
        e[i][0].call(p.exports, function (r) {
          var n = e[i][1][r];
          return o(n || r);
        }, p, p.exports, r, e, n, t);
      }

      return n[i].exports;
    }

    for (var u = "function" == typeof require && require, i = 0; i < t.length; i++) o(t[i]);

    return o;
  }

  return r;
})()({
  1: [function (require, module, exports) {
    const foo = 'bar';
  }, {}]
}, {}, [1]);"
`;

exports[`plugin forwards the options 1`] = `
"(function () {
  function r(e, n, t) {
    function o(i, f) {
      if (!n[i]) {
        if (!e[i]) {
          var c = "function" == typeof require && require;
          if (!f && c) return c(i, !0);
          if (u) return u(i, !0);
          var a = new Error("Cannot find module '" + i + "'");
          throw a.code = "MODULE_NOT_FOUND", a;
        }

        var p = n[i] = {
          exports: {}
        };
        e[i][0].call(p.exports, function (r) {
          var n = e[i][1][r];
          return o(n || r);
        }, p, p.exports, r, e, n, t);
      }

      return n[i].exports;
    }

    for (var u = "function" == typeof require && require, i = 0; i < t.length; i++) o(t[i]);

    return o;
  }

  return r;
})()({
  1: [function (require, module, exports) {
    // foo bar

    /* baz qux */
    const foo = 'bar';
  }, {}]
}, {}, [1]);"
`;

exports[`plugin generates a source map 1`] = `
"(function () {
  function r(e, n, t) {
    function o(i, f) {
      if (!n[i]) {
        if (!e[i]) {
          var c = "function" == typeof require && require;
          if (!f && c) return c(i, !0);
          if (u) return u(i, !0);
          var a = new Error("Cannot find module '" + i + "'");
          throw a.code = "MODULE_NOT_FOUND", a;
        }

        var p = n[i] = {
          exports: {}
        };
        e[i][0].call(p.exports, function (r) {
          var n = e[i][1][r];
          return o(n || r);
        }, p, p.exports, r, e, n, t);
      }

      return n[i].exports;
    }

    for (var u = "function" == typeof require && require, i = 0; i < t.length; i++) o(t[i]);

    return o;
  }

  return r;
})()({
  1: [function (require, module, exports) {
    module.exports.onRpcRequest = ({
      request
    }) => {
      console.log("Hello, world!");
      const {
        method,
        id
      } = request;
      return method + id;
    };
  }, {}]
}, {}, [1]);
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJtYXBwaW5ncyI6IkFBQUE7RUFBQTtJQUFBO01BQUE7UUFBQTtVQUFBO1VBQUE7VUFBQTtVQUFBO1VBQUE7UUFBQTs7UUFBQTtVQUFBQTtRQUFBO1FBQUFDO1VBQUE7VUFBQTtRQUFBO01BQUE7O01BQUE7SUFBQTs7SUFBQTs7SUFBQTtFQUFBOztFQUFBO0FBQUE7RUFBQTtJQ0NBQztNQUFBQztJQUFBO01BQ0FDO01BRUE7UUFBQUM7UUFBQUM7TUFBQTtNQUNBO0lBQ0EsQ0FMQTtHRERBO0FBQUEiLCJuYW1lcyI6WyJleHBvcnRzIiwiZSIsIm1vZHVsZSIsInJlcXVlc3QiLCJjb25zb2xlIiwibWV0aG9kIiwiaWQiXSwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsiLi4vLi4vbm9kZV9tb2R1bGVzL2Jyb3dzZXItcGFjay9fcHJlbHVkZS5qcyIsIl9zdHJlYW1fMC5qcyJdLCJzb3VyY2VzQ29udGVudCI6WyIoZnVuY3Rpb24oKXtmdW5jdGlvbiByKGUsbix0KXtmdW5jdGlvbiBvKGksZil7aWYoIW5baV0pe2lmKCFlW2ldKXt2YXIgYz1cImZ1bmN0aW9uXCI9PXR5cGVvZiByZXF1aXJlJiZyZXF1aXJlO2lmKCFmJiZjKXJldHVybiBjKGksITApO2lmKHUpcmV0dXJuIHUoaSwhMCk7dmFyIGE9bmV3IEVycm9yKFwiQ2Fubm90IGZpbmQgbW9kdWxlICdcIitpK1wiJ1wiKTt0aHJvdyBhLmNvZGU9XCJNT0RVTEVfTk9UX0ZPVU5EXCIsYX12YXIgcD1uW2ldPXtleHBvcnRzOnt9fTtlW2ldWzBdLmNhbGwocC5leHBvcnRzLGZ1bmN0aW9uKHIpe3ZhciBuPWVbaV1bMV1bcl07cmV0dXJuIG8obnx8cil9LHAscC5leHBvcnRzLHIsZSxuLHQpfXJldHVybiBuW2ldLmV4cG9ydHN9Zm9yKHZhciB1PVwiZnVuY3Rpb25cIj09dHlwZW9mIHJlcXVpcmUmJnJlcXVpcmUsaT0wO2k8dC5sZW5ndGg7aSsrKW8odFtpXSk7cmV0dXJuIG99cmV0dXJuIHJ9KSgpIiwiXG4gIG1vZHVsZS5leHBvcnRzLm9uUnBjUmVxdWVzdCA9ICh7IHJlcXVlc3QgfSkgPT4ge1xuICAgIGNvbnNvbGUubG9nKFwiSGVsbG8sIHdvcmxkIVwiKTtcblxuICAgIGNvbnN0IHsgbWV0aG9kLCBpZCB9ID0gcmVxdWVzdDtcbiAgICByZXR1cm4gbWV0aG9kICsgaWQ7XG4gIH07XG4iXX0="
`;

exports[`plugin processes files using Browserify 1`] = `
"(function () {
  function r(e, n, t) {
    function o(i, f) {
      if (!n[i]) {
        if (!e[i]) {
          var c = "function" == typeof require && require;
          if (!f && c) return c(i, !0);
          if (u) return u(i, !0);
          var a = new Error("Cannot find module '" + i + "'");
          throw a.code = "MODULE_NOT_FOUND", a;
        }

        var p = n[i] = {
          exports: {}
        };
        e[i][0].call(p.exports, function (r) {
          var n = e[i][1][r];
          return o(n || r);
        }, p, p.exports, r, e, n, t);
      }

      return n[i].exports;
    }

    for (var u = "function" == typeof require && require, i = 0; i < t.length; i++) o(t[i]);

    return o;
  }

  return r;
})()({
  1: [function (require, module, exports) {
    module.exports.onRpcRequest = ({
      request
    }) => {
      console.log("Hello, world!");
      const {
        method,
        id
      } = request;
      return method + id;
    };
  }, {}]
}, {}, [1]);"
`;
